grovel/grovel.lisp

```diff
@@ -179,68 +195,140 @@ int main(int argc, char**argv) {
   ;; found.
   (intern (symbol-name (car form)) '#:cffi-grovel))
 
-(defvar *header-forms* '(c include define flag typedef))
+(defvar *header-forms* '(c include-local include define flag typedef))
 
 (defun header-form-p (form)
   (member (form-kind form) *header-forms*))
 
-(defun generate-c-file (input-file output-defaults)
-  (nest
-   (with-standard-io-syntax)
-   (let ((c-file (make-c-file-name output-defaults "__grovel"))
-         (*print-readably* nil)
-         (*print-escape* t)))
-   (with-open-file (out c-file :direction :output :if-exists :supersede))
-   (with-open-file (in input-file :direction :input))
-   (flet ((read-forms (s)
-            (do ((forms ())
-                 (form (read s nil nil) (read s nil nil)))
-                ((null form) (nreverse forms))
-              (labels
-                  ((process-form (f)
-                     (case (form-kind f)
-                       (flag (warn "Groveler clause FLAG is deprecated, use CC-FLAGS instead.")))
-                     (case (form-kind f)
-                       (in-package
-                        (setf *package* (find-package (second f)))
-                        (push f forms))
-                       (progn
-                         ;; flatten progn forms
-                         (mapc #'process-form (rest f)))
-                       (t (push f forms)))))
-                (process-form form))))))
-   (let* ((forms (read-forms in))
-          (header-forms (remove-if-not #'header-form-p forms))
-          (body-forms (remove-if #'header-form-p forms)))
-     (write-string *header* out)
-     (dolist (form header-forms)
-       (process-grovel-form out form))
-     (write-string *prologue* out)
-     (dolist (form body-forms)
-       (process-grovel-form out form))
-     (write-string *postscript* out)
-     c-file)))
+(defun generate-c-file (c-file forms)
+  (with-standard-io-syntax
+    (let ((*print-readably* nil)
+          (*print-escape* t))
+      (with-open-file (out c-file :direction :output :if-exists :supersede)
+        (let* ((header-forms (remove-if-not #'header-form-p forms))
+               (body-forms (remove-if #'header-form-p forms)))
+          (write-string *header* out)
+          (dolist (form header-forms)
+            (process-grovel-form out form))
+          (write-string *prologue* out)
+          (dolist (form body-forms)
+            (process-grovel-form out form))
+          (write-string *postscript* out)
+          c-file)))))
```

Comment:
The old version was hard to read as a newcomer because of nest. It is also much smaller now as the forms are read earlier in the process and passed to this function.